File: vestigraph/web/static/skills.js
// Skill UI owns its dialogs and drops responses after close, navigation or project changes.
export function setupSkills({api,t,getDocument,getSelected,getProjects,getProject}) {
  const $ = id => document.getElementById(id);
  const createDialog=$("skill-create-dialog"), library=$("skill-library-dialog");
  let generation=0, did=null, pid=null, revision=0, cursor=null, before=null, current=null;
  let options=null, busy=false, dirty=false, timer=null, latestRevision=0, expanded=false, polling=false;
  const enc=encodeURIComponent;
  const status=(text,create=false)=>{$(create?"skill-create-status":"skill-library-status").textContent=text;};
  function buttons() {
    for(const id of ["skill-create-submit","skill-save","skill-publish","skill-validate","skill-generate","skill-previous","skill-latest","skill-library-refresh","skill-list-more","skill-more-versions"]) $(id).disabled=busy;
    $("skill-body").disabled=busy;
    $("skill-project").disabled=busy;
    if(current) {
      const old=current.revision!==latestRevision;
      $("skill-save").disabled=busy||old;
      $("skill-publish").disabled=busy||dirty||old||current.body_kind==="outline"||current.state==="awaiting_agent";
      $("skill-validate").disabled=busy||dirty||old||!options?.validators.length;
      $("skill-generate").disabled=busy||dirty||old||!options?.generators.length;
      $("skill-previous").disabled=busy||current.revision<=1;
      $("skill-latest").disabled=busy;
    }
  }
  async function run(fn,creation=false) {
    if(busy) return;
    const token=generation;busy=true;buttons();
    try {await fn(token);} catch(e) {if(token===generation) {status(e.message||String(e),creation);if(e.code==="SKILL_REVISION_CONFLICT")$("skill-latest").disabled=false;}}
    finally {if(token===generation){busy=false;buttons();}}
  }
  function close() {generation++;busy=false;clearInterval(timer);timer=null;}
  createDialog.addEventListener("close",()=>{if(!library.open)close();});
  library.addEventListener("close",()=>{if(!createDialog.open)close();});
  $("skill-create-close").onclick=()=>createDialog.close();
  $("skill-library-close").onclick=()=>library.close();
  window.addEventListener("hashchange",()=>{if(createDialog.open)createDialog.close();if(library.open)library.close();});
  function providerOptions(id,values) {
    $(id).replaceChildren();
    for(const item of values) {const o=document.createElement("option");o.value=item.id;
      const key="skills.provider_"+item.id;const translated=t(key);
      o.textContent=translated===key ? item.label : translated;$(id).append(o);}
  }
  async function loadVersions(token,next=false) {
    const page=await api.checkpoints(did,{limit:100,cursor:next?cursor:null});
    if(token!==generation)return;
    if(!next) {revision=page.history_revision||0;$("skill-from").replaceChildren();$("skill-to").replaceChildren();}
    if((page.history_revision||0)!==revision)throw new Error(t("skills.history_changed"));
    for(const cp of page.items) for(const id of ["skill-from","skill-to"]) {
      const o=document.createElement("option");o.value=cp.id;o.textContent=cp.title||cp.filename;$(id).append(o);
    }
    cursor=page.next_cursor;$("skill-more-versions").hidden=!cursor;
    if(!next) {
      const select=$("skill-to"), selected=getSelected();
      if([...select.options].some(o=>o.value===selected))select.value=selected;
      $("skill-from").selectedIndex=Math.min(select.selectedIndex+1,$("skill-from").options.length-1);
    }
  }
  $("skill-refine").onclick=()=>{
    if(!getDocument())return;
    did=getDocument();generation++;createDialog.showModal();$("skill-create-form").reset();status("",true);
    run(async token=>{
      if(!api.capabilities.skill_library)throw new Error(t("skills.old_service"));
      options=await api.get("/skill-options");if(token!==generation)return;
      providerOptions("skill-processing",[{id:"agent",label:t("skills.external_agent")},...options.generators]);
      await loadVersions(token);
      if(token===generation&&$("skill-to").options.length<2)status(t("skills.need_two"),true);
    },true);
  };
  $("skill-more-versions").onclick=()=>run(token=>loadVersions(token,true),true);
  $("skill-create-form").onsubmit=e=>{
    e.preventDefault();
    run(async token=>{
      const intent={};for(const key of ["title","goal","rationale","applicability","parameters","success_criteria","domain"]) intent[key]=$("skill-"+key).value;
      const mode=$("skill-processing").value;
      const item=await api.post("/documents/"+enc(did)+"/skills",{intent,source:{provider:"history-window",selection:{from_id:$("skill-from").value,to_id:$("skill-to").value,history_revision:revision}}});
      if(token!==generation)return;
      createDialog.close();
      await openLibrary(item.id,item.project_id);
      if(mode!=="agent"&&library.open&&current?.id===item.id){$("skill-generator").value=mode;await run(tok=>job("generate","skill-generator",tok));}
    },true);
  };
  async function list(token,next=false) {
    const result=await api.get("/projects/"+enc(pid)+"/skills"+(next&&before?"?before="+before:""));
    if(token!==generation)return;
    if(!next){$("skill-cards").replaceChildren();expanded=false;}else expanded=true;
    for(const item of result.items) {
      const card=document.createElement("article");card.className="card";
      const b=document.createElement("button");b.type="button";b.textContent=item.title;b.dataset.skillId=item.id;
      b.onclick=()=>run(async tok=>{if(dirty)throw new Error(t("skills.unsaved"));await load(item.id,tok);});
      const p=document.createElement("p");p.className="muted";p.textContent=t("skills.state_"+item.state)+" · v"+item.revision;
      card.append(b,p);$("skill-cards").append(card);
    }
    before=result.next_before;$("skill-list-more").hidden=!before;
    if(!$("skill-cards").children.length)status(t("skills.empty"));
  }
  function render() {
    $("skill-editor").hidden=!current;if(!current)return;
    $("skill-editor-title").textContent=current.intent.title;
    $("skill-state").textContent=t("skills.state_"+current.state)+" · v"+current.revision;
    $("skill-body").value=current.body;dirty=false;
    const source=current.sources.data;
    $("skill-source").textContent=(source.document_name||current.document_id)+" · "+t("skills.source_count",{count:source.artifacts?.length||0});
    const q=new URLSearchParams({doc:current.document_id});
    if(source.selection?.to_id)q.set("version",source.selection.to_id);
    $("skill-source-link").href="#"+q.toString();
    $("skill-evidence").textContent=JSON.stringify({explanation:current.intent,evidence:source},null,2);
    $("skill-files").textContent=t("skills.files",{count:Object.keys(current.files||{}).length});
    const verification=current.verification;
    $("skill-verification").textContent=t("skills.verification_"+verification.status)+
      (verification.reports.length ? "\n"+verification.reports.map(r=>[r.scope,r.status,r.message].filter(Boolean).join(" · ")).join("\n"):"");
    exportLink();buttons();
  }
  function exportLink() {if(current)$("skill-export").href="/api/v1/skills/"+enc(current.id)+"/export?format="+enc($("skill-exporter").value||"agent-skill")+"&revision="+current.revision;}
  async function load(id,token,version=null) {
    const item=await api.get("/skills/"+enc(id)+(version?"?revision="+version:""));
    if(token!==generation)return;
    current=item;if(!version)latestRevision=item.revision;render();
  }
  async function openLibrary(id=null,project=null) {
    generation++;busy=false;dirty=false;current=null;$("skill-editor").hidden=true;status("");
    if(!library.open)library.showModal();
    pid=project||getProject()||getProjects()[0]?.id;
    $("skill-project").replaceChildren();
    for(const p of getProjects()){const o=document.createElement("option");o.value=p.id;o.textContent=p.name;$("skill-project").append(o);}
    $("skill-project").value=pid||"";
    await run(async token=>{
      if(!api.capabilities.skill_library)throw new Error(t("skills.old_service"));
      options=await api.get("/skill-options");if(token!==generation)return;
      providerOptions("skill-generator",options.generators);providerOptions("skill-validator",options.validators);providerOptions("skill-exporter",options.exporters);
      if(!pid)throw new Error(t("skills.no_project"));
      await list(token);if(id&&token===generation)await load(id,token);
    });
    clearInterval(timer);
    if(!library.open||!api.capabilities.skill_library||!pid)return;
    timer=setInterval(async()=>{
      if(!library.open||busy||dirty||polling)return;
      const token=generation,id=current?.id,version=current?.revision;
      polling=true;
      try {
        const item=id&&version===latestRevision ? await api.get("/skills/"+enc(id)) : null;
        if(token!==generation||busy||dirty||document.activeElement===$("skill-body"))return;
        if(item&&current?.id===id&&current.revision===version&&item.revision!==version){
          current=item;latestRevision=item.revision;render();
        }
        if(!expanded&&!library.contains(document.activeElement))await list(token);
      } catch(e) {if(token===generation)status(e.message||String(e));}
      finally {polling=false;}
    },5000);
  }
  $("skills-open").onclick=()=>openLibrary();
  $("skill-project").onchange=()=>{generation++;busy=false;current=null;dirty=false;pid=$("skill-project").value;$("skill-editor").hidden=true;run(token=>list(token));};
  $("skill-library-refresh").onclick=()=>run(async token=>{if(dirty)throw new Error(t("skills.unsaved"));await list(token);if(current)await load(current.id,token);});
  $("skill-list-more").onclick=()=>run(token=>list(token,true));
  $("skill-body").oninput=()=>{dirty=true;buttons();};
  $("skill-save").onclick=()=>run(async token=>{
    const item=await api.put("/skills/"+enc(current.id),{expected_revision:current.revision,body:$("skill-body").value,files:current.files,actor:"user"});
    if(token!==generation)return;current=item;latestRevision=item.revision;render();await list(token);status(t("skills.saved"));
  });
  $("skill-publish").onclick=()=>run(async token=>{
    const item=await api.post("/skills/"+enc(current.id)+"/publish",{expected_revision:current.revision});
    if(token!==generation)return;current=item;latestRevision=item.revision;render();await list(token);status(t("skills.published"));
  });
  async function job(kind,select,token) {
    const id=current.id;
    const accepted=await api.post("/skills/"+enc(id)+"/"+kind,{expected_revision:current.revision,provider:$(select).value});
    for(;;) {
      if(token!==generation)return;
      const j=await api.job(accepted.job_id);if(token!==generation)return;
      if(j.status==="succeeded"){await load(id,token);await list(token);status(t("skills.task_done"));return;}
      if(!["queued","running"].includes(j.status))throw new Error(j.error?.message||t("skills.task_failed"));
      status(t("skills.task_running"));await new Promise(resolve=>setTimeout(resolve,500));
    }
  }
  $("skill-validate").onclick=()=>run(token=>job("validate","skill-validator",token));
  $("skill-generate").onclick=()=>run(token=>job("generate","skill-generator",token));
  $("skill-exporter").onchange=exportLink;
  $("skill-previous").onclick=()=>run(async token=>{if(dirty)throw new Error(t("skills.unsaved"));await load(current.id,token,current.revision-1);});
  $("skill-latest").onclick=()=>run(async token=>{if(dirty&&!window.confirm(t("skills.discard_changes")))return;await load(current.id,token);});
  $("skill-copy-task").onclick=()=>run(async token=>{
    const text="Complete the Vestigraph skill request "+current.id+" using the local KLink MCP tools. Start with klink.status, discover the vestigraph domain, then call vestigraph.skill with skill_id="+current.id+" and follow its next_action. Submit a draft with expected_revision="+current.revision+" after writing instructions from the frozen evidence. Preserve evidence, user intent and inference distinctions; do not publish, upload, execute attachments or replay against the live design.";
    await navigator.clipboard.writeText(text);if(token===generation)status(t("skills.copied"));
  });
}

File: vestigraph/service/skills.py
"""Skill application service shared by the web UI and registered agent tools."""
import copy
import re
from ..vesti_skills.catalog import SkillCatalog, encoded
from ..vesti_skills.contracts import default_services
from ..vesti_skills.draft import EvidenceDraft
from .catalog import fingerprint
from .errors import bad_request, conflict


def intent_of(value):
    allowed = {"title", "goal", "rationale", "applicability", "parameters", "success_criteria", "domain"}
    if not isinstance(value, dict) or set(value) - allowed:
        raise bad_request("Invalid skill explanation fields.")
    result = {}
    for key in allowed:
        text = value.get(key, "")
        limit = 200 if key in ("title", "domain") else 4000
        if not isinstance(text, str) or len(text) > limit:
            raise bad_request("Invalid or oversized skill explanation: " + key)
        result[key] = text.strip()
    if not result["title"] or not result["goal"]:
        raise bad_request("Skill name and goal are required.")
    return result


def content_of(body, files):
    if not isinstance(body, str) or not body.strip() or len(body.encode("utf-8")) > 65536:
        raise bad_request("Skill instructions must be nonempty and at most 64 KiB.")
    if not isinstance(files, dict) or len(files) > 20:
        raise bad_request("A skill may have at most 20 text supporting files.")
    used = set()
    for name, text in files.items():
        if (not isinstance(name, str) or len(name) > 160
                or not re.fullmatch(r"(scripts|references|assets)/[a-zA-Z0-9_./-]+", name)
                or any(part in ("", ".", "..") for part in name.split("/"))
                or name.lower() == "references/vestigraph.json"
                or name.lower() in used or not isinstance(text, str)):
            raise bad_request("Supporting files need unique safe relative paths under scripts, references or assets.")
        used.add(name.lower())
    if len(encoded(files).encode()) > 128 * 1024:
        raise bad_request("Supporting text files exceed 128 KiB.")
    return body.strip(), dict(files)


class Skills:
    def __init__(self, app, services=None):
        self.app = app
        self.services = services or default_services()
        self.catalog = SkillCatalog(app.catalog)
        app.runner.register("skill_generate", "heavy", self._generate)
        app.runner.register("skill_validate", "heavy", self._validate)

    def options(self):
        return {"schema_version": 1, "agent_handoff": True, **self.services.describe()}

    def list(self, project_id, before=None, state=None, limit=30):
        self.app.catalog.get_project(project_id)
        return self.catalog.list(project_id, before, state, limit)

    def get(self, sid, revision=None):
        return self.catalog.get(sid, revision)

    def create(self, document_id, intent, source, request_key=None):
        document = self.app.catalog.get_document(document_id)
        intent = intent_of(intent)
        if (not isinstance(source, dict) or set(source) != {"provider", "selection"}
                or not isinstance(source["provider"],str) or not isinstance(source["selection"],dict)):
            raise bad_request("A source provider and selection are required.")
        provider = self.services.sources.get(source["provider"])
        if provider is None:
            raise bad_request("The selected evidence source provider is not registered.")
        evidence = provider.collect(self.app, document, source["selection"])
        if not isinstance(evidence, dict):
            raise bad_request("Evidence provider returned an invalid bundle.")
        # Enforce a common provenance boundary even for domain-specific collectors.
        encoded(evidence)
        bundle = {"provider": provider.id, "selection": copy.deepcopy(source["selection"]),
                  "document_id": document_id, "data": evidence, "digest": fingerprint(evidence)}
        outline = EvidenceDraft().generate(evidence, intent)
        body, files = content_of(outline["body"], {})
        payload = {"intent": intent, "sources": bundle, "body": body, "files": files,
                   "state": "awaiting_agent", "body_kind": "outline",
                   "generation": {"provider": None, "actor": "user-request", "mode": "agent_handoff"},
                   "verification": {"status": "not_verified", "reports": []}}
        return self.catalog.create(document["project_id"], document_id, payload, request_key)

    def revise(self, sid, expected_revision, body, files=None, intent=None, actor="user", validation_note="", *, check_structure=False):
        body, files = content_of(body, {} if files is None else files)
        if not isinstance(actor, str) or not 1 <= len(actor) <= 100:
            raise bad_request("Invalid skill author.")
        if not isinstance(validation_note, str) or len(validation_note) > 4000:
            raise bad_request("Validation note must be at most 4000 characters.")
        updated_intent = intent_of(intent) if intent is not None else None
        structure_report = None
        if check_structure:
            from ..vesti_skills.draft import ContractValidator
            structure_report = ContractValidator().validate({"body": body, "files": files})
            if structure_report["status"] != "passed":
                raise bad_request("Skill structure check failed.", "Read the selected request and submit nonempty instructions.")
        def update(item):
            item.update(body=body, files=files, state="draft", body_kind="instructions")
            if updated_intent is not None:
                item["intent"] = updated_intent
            item["generation"] = {**item["generation"], "actor": actor, "mode": "submitted"}
            item["verification"] = {"status": "not_verified", "reports": []}
            if validation_note.strip():
                item["verification"] = {"status": "reported", "reports": [{
                    "status": "reported", "scope": "author_statement", "actor": actor,
                    "message": validation_note.strip(), "content_digest": fingerprint({"body":body,"files":files}),
                    "source_digest": item["sources"]["digest"]}]}
            if structure_report is not None:
                report = dict(structure_report, provider="skill-contract", revision=expected_revision + 1,
                              content_digest=fingerprint({"body": body, "files": files}),
                              source_digest=item["sources"]["digest"])
                item["verification"] = {"status": "checked", "reports": [report, *item["verification"]["reports"]]}
        return self.catalog.update(sid, expected_revision, update)

    def publish(self, sid, expected_revision):
        def update(item):
            if item["state"] == "awaiting_agent" or item["body_kind"] == "outline":
                raise bad_request("Complete and save the skill instructions before publishing.")
            item["state"] = "published"
        return self.catalog.update(sid, expected_revision, update)

    def request(self, sid, expected_revision, provider_id, kind, request_key=None):
        item = self.get(sid)
        if item["revision"] != expected_revision:
            raise conflict("SKILL_REVISION_CONFLICT", "Reload the latest skill before starting this task.")
        providers = self.services.generators if kind == "skill_generate" else self.services.validators
        if provider_id not in providers:
            raise bad_request("The selected skill provider is not registered.")
        job, _ = self.app.runner.submit(item["project_id"], kind, "skill", sid,
            {"revision":expected_revision, "provider":provider_id}, request_key=request_key)
        return job

    def _generate(self, job):
        current = self.get(job["target_id"], job["payload"]["revision"])
        provider = self.services.generators[job["payload"]["provider"]]
        result = provider.generate(copy.deepcopy(current["sources"]["data"]), copy.deepcopy(current["intent"]))
        body, files = content_of(result.get("body"), result.get("files", {}))
        def update(item):
            item.update(body=body, files=files, state="draft",
                        body_kind="outline" if result.get("kind") == "outline" else "instructions",
                        generation={"provider":provider.id,"actor":provider.id,"mode":"provider"},
                        verification={"status":"not_verified","reports":[]})
        item = self.catalog.update(current["id"], current["revision"], update)
        return {"skill_id":item["id"],"revision":item["revision"]}, None

    def _validate(self, job):
        current = self.get(job["target_id"], job["payload"]["revision"])
        provider = self.services.validators[job["payload"]["provider"]]
        report = provider.validate(copy.deepcopy(current))
        if not isinstance(report, dict) or report.get("status") not in ("passed","failed","not_verified") or not isinstance(report.get("scope"), str):
            raise bad_request("Validator must report its status and validation scope.")
        encoded(report)
        report.update(provider=provider.id, revision=current["revision"],
                      content_digest=fingerprint({"body":current["body"],"files":current["files"]}),
                      source_digest=current["sources"]["digest"])
        def update(item):
            item["verification"] = {"status": "checked", "reports": [report]}
        item = self.catalog.update(current["id"],current["revision"],update)
        return {"skill_id":item["id"],"revision":item["revision"],"report":report}, None

    def export(self, sid, exporter_id, revision=None):
        provider = self.services.exporters.get(exporter_id)
        if provider is None:
            raise bad_request("The selected skill exporter is not registered.")
        return provider.export(self.get(sid,revision))
